wiki/DEFINITION_OF_DONE_CRITERIA.md

```diff
@@ -0,0 +1,99 @@
+- **Testing:** The goal for automated testing is to increase the code coverage as much as possible in order to ease time spent fixing bugs and developing new features. Additionally, a very important factor for a tool like `cuml` is to provide testing with multitple datasets that really stress the mathematical behavior of the algorithms, even if they don't increase code coverage! A comprehensive set of tests lowers the possibility for regressions and the introduction of bugs as the code evolves between versions. This covers both correctness & performance. 
+- **Documentation:** User-facing documentation should be complete and descriptive. Developer-facing documentation should be used for constructs which are complex and/or not immediately obvious. 
+
+### C++
+
+#### Design
+
+- Existing prims are used wherever possible
+- Array inputs and outputs to algorithms are accepted on device
+- New prims created wherever there is potential for reuse across different algorithms or prims
+- User-facing API is stateless and follows the plain-old data (POD) design paradigm
+- Public API contains a C-Wrapper around the stateless API
+- Public API contains an Scikit-learn-like stateful wrapper around the stateless API
+
+#### Testing
+
+- Prims: GTests with different inputs
+- Algorithms: End-to-end GTests with different inputs and different datasets
+
+#### Documentation
+
+- Complete and comprehensive Doxygen strings explaining the public API, restrictions, and gotchas. Any array parameters should also note whether the underlying memory is host or device. 
+
+
+### Python
+
+#### Design
+
+- Python class is as "near drop-in replacement" for Scikit-learn (or relevant industry standard) API as possible
+    - It is recommended to open an initial PR with the API design if there are going to be significant differences with reference APIs, or lack of a reference API, to have a discussion about it. 
+- Python class is pickleable
+- APIs use `input_to_dev_array` to accept flexible inputs and check their datatypes
+
+#### Testing 
+
+- Pytests for wrapper functionality against Scikit-learn using relevant datasets
+- Stress tests against reasonable inputs (e.g short-wide, tall-narrow, different numerical precision)
+- Pytests for pickle capability
+- Pytests to evaluate correctness against Scikit-learn on a variety of datasets
+- Add algorithm to benchmarks packages
+
+#### Documentation
+
+- Complete and comprehensive Pydoc strings explaining public API, restrictions, and gotchas
+- Docstrings include references to the underlying algorithm (e.g paper or Scikit-learn algorithm being implemented or a description of the algorithm used if nonstandard).
+
+
+## Algorithm Status
+
+Sometimes it is possible that only some of these requirements are satisfied before an algorithm gets released. We assign a status of `bronze`, `silver`, or `gold` to denote the amount of work necessary to provide the most robust, performant, and maintainable algorithm possible.
```

Comment:
Should we go through and backfill these classifications for existing algos at the beginning of 0.10? Are you thinking we should make a list on the wiki or embed in the docstring for each algo? (I'm kind of partial to including it in each algo so it's immediately obvious to someone using the code, but I could be wrong.)

Reply:
I agree with @JohnZed. Such a classification would in-turn also help us in prioritizing the work.

How about this?
We can create a checklist based on the items described in this doc.
Depending on how many items have been checked-off from this list, we can attach a number to it.
Cumulative number could then be used to decide this classification.